src/lib/statistics/chunk_statistics/histograms/abstract_histogram.hpp

```diff
@@ -0,0 +1,214 @@
+   * This is equal to the length of the segment during creation, without null values.
+   */
+  virtual HistogramCountType total_count() const = 0;
+
+  /**
+   * Returns the number of distinct values represented in the histogram.
+   * This is equal to the number of distinct values in the segment during creation.
+   */
+  virtual HistogramCountType total_distinct_count() const = 0;
+
+ protected:
+  /**
+   * Returns a list of pairs of distinct values and their respective number of occurrences in a given segment.
+   * The list is sorted by distinct value from lowest to highest.
+   */
+  static std::vector<std::pair<T, HistogramCountType>> _value_counts_in_segment(
+      const std::shared_ptr<const BaseSegment>& segment);
+
+  /**
+   * Calculates the estimated cardinality for predicate types supported by all data types.
+   */
+  float _estimate_cardinality(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                              const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Makes pruning decisions for predicate types supported by all data types.
+   */
+  bool _can_prune(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                  const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Given a value return the next representable value.
+   * This method is a wrapper for the functions in histogram_utils.
+   */
+  T _get_next_value(const T value) const;
+
+  /**
+   * Given a value return its numerical representation.
+   * This method is a wrapper for the functions in histogram_utils.
+   */
+  uint64_t _convert_string_to_number_representation(const std::string& value) const;
+
+  /**
+   * Given a numerical representation of a string return the string.
+   * This method is a wrapper for the functions in histogram_utils.
+   */
+  std::string _convert_number_representation_to_string(const uint64_t value) const;
+
+  /**
+   * Returns the share of values in a bin that are lower than `value`.
+   * This method is specialized for strings.
+   */
+  float _bin_share(const BinID bin_id, const T value) const;
+
+  /**
+   * Returns the width of a bin.
+   * This method is specialized for strings to return a numerical width.
+   */
+  virtual HistogramWidthType _bin_width(const BinID index) const;
+
+  /**
+   * Returns the id of the bin that holds the given `value`.
+   * Returns INVALID_BIN_ID if `value` does not belong to any bin.
+   */
+  virtual BinID _bin_for_value(const T value) const = 0;
+
+  /**
+   * Returns the id of the bin after the one that holds the given `value`.
+   * If `value` does not belong to any bin but is smaller than max(), it is in a gap.
+   * In that case return the bin right after the gap.
+   * If the bin that holds the value is the last bin or it is greater than max, return INVALID_BIN_ID.
+   */
+  virtual BinID _upper_bound_for_value(const T value) const = 0;
```

Comment:
"upper bound" usually has a different meaning, see the STL algorithms.

Reply:
Yes, I agree. What do you think about `_bin_after_value()`?

Reply:
yepp. Or `next_bin`?

Reply:
Works for me, changed.